File: haskell/src/Solutions/Euler277.hs
module Solutions.Euler277 where
import Data.Maybe
import Data.Ratio
import Data.List

data Poly = Poly Rational Rational -- Poly a b === ax + b
infixl 7 ***
(***) :: Rational -> Poly -> Poly
n *** (Poly a b) = Poly (n * a) (n * b)
infixl 6 +++
(+++) :: Rational -> Poly -> Poly
n +++ (Poly a b) = Poly a (n + b)
eval :: Poly -> Rational -> Rational
eval (Poly a b) r = a * r + b

decipher :: String -> Poly
decipher "" = Poly 1 0
decipher ('D':t) = 3 *** decipher t
decipher ('U':t) = (1 % 4) *** ((-2) +++ 3 *** decipher t)
decipher ('d':t) = (1 % 2) *** (1 +++ 3 *** decipher t)

msg = "UDDDUdddDDUDDddDdDddDDUDDdUUDd"
p = decipher msg

leastX = fromJust $ find (\x -> denominator (eval p x) == 1) [1..]
leastA = eval p leastX
leastA' = numerator $ fromJust $ find (>10^15) [leastA, leastA + 3 ^ length msg..]

euler277 :: IO String
euler277 = return $ show leastA'

File: rust/src/solutions/euler277.rs
use num::{rational::Ratio, One, Zero};
use std::ops::{Add, Mul};

type R = Ratio<i128>;

fn ratio(n: i128, d: i128) -> R {
    Ratio::new(n, d)
}

struct Poly(R, R); // (Poly a b)(x) == a x + b

impl<'a> Add<&'a R> for &'a Poly {
    type Output = Poly;
    fn add(self, rhs: &'a R) -> Self::Output {
        Poly(self.0.clone(), rhs + &self.1)
    }
}

impl<'a> Mul<&'a R> for &'a Poly {
    type Output = Poly;
    fn mul(self, rhs: &'a R) -> Self::Output {
        Poly(rhs * &self.0, rhs * &self.1)
    }
}

fn eval(poly: &Poly, at: &R) -> R {
    &poly.0 * at + &poly.1
}

fn decipher(chars: &[char]) -> Poly {
    if chars.is_empty() {
        Poly(R::one(), R::zero())
    } else {
        let d3 = decipher(&chars[1..]).mul(&ratio(3, 1));
        match chars[0] {
            'D' => d3,
            'U' => d3.add(&ratio(-2, 1)).mul(&ratio(1, 4)),
            _ => d3.add(&ratio(1, 1)).mul(&ratio(1, 2)),
        }
    }
}

pub fn euler277() -> String {
    let msg = "UDDDUdddDDUDDddDdDddDDUDDdUUDd";
    let the_poly = decipher(&msg.chars().collect::<Vec<char>>()[..]);
    let least_x = (1..)
        .map(|x| ratio(x, 1))
        .find(|x| eval(&the_poly, x).denom() == &1)
        .unwrap();
    let least_a = *eval(&the_poly, &least_x).numer();
    (least_a as usize..)
        .step_by(3usize.pow(msg.len() as u32))
        .find(|&n| n > 10usize.pow(15))
        .unwrap()
        .to_string()
}
